### PR TITLE
DOC: update pydata-sphinx theme

### LINKED ISSUE #16944: BUG: scrollbar not visible in docs
### Describe your issue.

The browser's scrollbar on the right is not visible when it is toggled by mouseover. I do not see the same behaviour for example in the NumPy docs which also use the pydata theme, so probably related to SciPy itself. To reproduce, just go to the docs page and hover over your browser's scrollbar.

Browser: Chrome

Scrollbar not toggled:

image

Scrollbar toggled:

image

### Reproducing Code Example

```python
not applicable
```

### Error message

```shell
not applicable
```

### SciPy/NumPy/Python version information

not applicable

### LINKED ISSUE #18535: DOC: Dev branch docs render Dev TOC while viewing API Reference
### Issue with current documentation:

When viewing https://scipy.github.io/devdocs/reference/index.html I see the table of contents for the developer guide, not the reference table of contents.

This is not an issue with the stable builds.

The issue persists after clearing the browser cache, so I assume that's unrelated.

### Idea or request for content:

It would be nice if the correct table of contents was displayed. It's curious that this isn't an issue for stable, so hopefully isn't related to how hosting is handled.

### Additional context (e.g. screenshots, GIFs)

https://scipy.github.io/devdocs/reference/index.html looks like:

image

https://docs.scipy.org/doc/scipy/reference/index.html looks like

image

### PR DESCRIPTION
**Pull request by tupui**
This is a provision for the next release of the PyData Sphinx Theme which should happen in the following days (RC2 already up).

This will be needed if the doc is build after 0.10 is released otherwise there will be no search bar. The new search bar is a way better experience. Overall the new version provides a lot of improvements.

### CODE REVIEW

**Comment by mdhaber** [comment]
@tupui tried to resolve merge conflicts. Hope I did it correctly. (Oops, probably shouldn't have skipped most CI. )

File: doc/source/conf.py
Comment (mdhaber) [inline doc/source/conf.py:185]:
```suggestion
  "navbar_end": ["search-field.html", "theme-switcher", 
                 "version-switcher", "navbar-icon-links"],
```

**Comment by tupui** [comment]
Thanks Matt for having a look and catching the lint issue 👍

**Review by choldgraf — APPROVED** [review]

**Comment by rgommers** [comment]
The new theme version is out, but I guess this is blocked until the build time issues reported in https://github.com/pydata/pydata-sphinx-theme/issues/889 are resolved?

**Comment by tupui** [comment]
@rgommers yes this is blocked until the build time issues are resolved. Unless we are fine with either removing the templates (not sure) or doubling build time (probably not). I've quickly pinned the version in #16922 so that it would not break or CI and not block them to release either.

**Comment by rgommers** [comment]
Thanks. Yes indeed, doubling build time would be really bad, it is already quite painful. Not sure about the templates either, needs trying.

**Comment by tupui** [comment]
Still an issue with 0.12. I will remove the templates so we can have a look at the output here. Besides the build time, the new versions are great improvements and it's sad we are missing out.

**Comment by tupui** [comment]
And again an issue with 0.13.

**Comment by h-vetinari** [comment]
Based on past discussion this seems unlikely to be solved for 1.11, but I'd be very happy to be proven wrong! :)

**Comment by tupui** [comment]
I am afraid you are correct... 😢

**Comment by tupui** [comment]
Still not there yet with the 0.14.1 release, but there seems to be an improvement. Now the build step finished and we are stuck at the upload.

**Comment by tylerjereddy** [comment]
It looks like they're closer to a root cause for the slowdown issue in https://github.com/pydata/pydata-sphinx-theme/issues/889, but it seems unlikely that will be fully solved before we branch in ~2 weeks, though a fix for `1.13.0` does seem like it might happen.

I suppose I could leave this to the last minute, but currently looks likely that I'd bump the milestone again.

**Comment by tylerjereddy** [comment]
I'm bumping the milestone after re-reading the "upstream" status--sounds like there's a modest chance we may be able to consider moving forward in a few months with `1.13.0` perhaps.

**Comment by tupui** [comment]
Great news from the team's team! 🚀  Once https://github.com/pydata/pydata-sphinx-theme/pull/1609 is in, we might be able to update. I will keep an eye and re-work this PR once they cut a new release.

**Comment by tupui** [comment]
Alright, we are golden! 

Thank you @drammock and  @choldgraf for tackling this 💪🎉

I will wait for the next official release and we can deploy this.

@tylerjereddy if the next point release can wait a bit, would be great to have this in.

**Comment by tupui** [comment]
I've seen, yay 😃 I will update the PR after work.

**Comment by h-vetinari** [comment]
Checking out the [page](https://output.circle-artifacts.com/output/job/be632ad3-7059-4c14-a7b3-3492cab1177f/artifacts/0/html/index.html) in the artefacts of the `build_docs` job, this still seems to have some issues:

image

The header looks pretty objectively worse than the current main page IMO:

image

While it's more of a minor stylistic nit, I kinda prefer the slightly warmer colour scheme we have now, and (if the search bar is permanently moved from left to top-right) the horizontal balance is off to me - the page is leaning left to my tastes.

Back to the new page, the following "Back to top" annoyingly does not fade away:

image

It disappears when scrolling _down_ slightly, but that's unintuitive IMO.

**Comment by tupui** [comment]
I will see if we have some custom CSS that would go against the template (I will check NumPy's config too).

The long term goal was to progressively remove as much customization as possible. The theme is supposed to have better defaults (more inclusive colours, better UI, etc.)

**Comment by h-vetinari** [comment]
> The long term goal was to progressively remove as much customization as possible.

Absolutely in favour of that and don't want to stand in the way. Though ideally we should try not to "regress" too much in terms of UI/UX while doing so.

**Comment by drammock** [comment]
In my local testing when fixing the theme I had tweaked your conf.py to put search back in the left sidebar. I'd recommend doing that here, and having a longer discussion / separate PR about which upstream layout and style changes you want to override and which you want to keep.

I think I also had tweaked the top navbar ... I'll be back at my desk later today and can tell you then how exactly I changed the conf.

-------- Original Message --------

**Comment by tupui** [comment]
What I need to really fix though is the number of levels for the API part.

We have a crazy "logic" with hidden tree (I tried once to refactor that madness but did not see it through) and we need the custom html for that. I had deactivated this bit during the debug to see if that fixed the build time issue.

**Comment by tupui** [comment]
Thanks @drammock 🙌 feel free to make a PR on my fork on this branch. Might be easier than sharing bits of code there.

File: doc_requirements.txt
Comment (lucascolley) [inline doc_requirements.txt:1]: should this be showing in the diff?

Comment (tupui) [reply doc_requirements.txt:1]: It was as leftover from a merge commit, it's removed now :)

**Comment by tupui** [comment]
I did the updates. Seems like we have the same state as on develop with the new version 😃 

There is still an issue with the sidebar TOC for the API reference page, but that's a known existing issue that we think is linked to how the website is hosted on CircleCI. Locally the TOC is there, style (not stacked) can be discussed later.

@drammock let us know if you have some quick fixes for 1. the navbar, 2. the back to top button and 3. TOC (though as I said, I think there is no change besides the style)

Screenshot 2024-01-04 at 21 54 06

**Comment by drammock** [comment]
@tupui I just sent a PR into this branch with a couple conf.py tweaks. However, for me locally the API pages don't just have a poorly formatted left sidebar TOC, they have *the wrong TOC* (the one from the "development" page). I see that here too: https://output.circle-artifacts.com/output/job/056b1ded-d18d-4e94-bdcb-cba697e7bc45/artifacts/0/html/reference/index.html

**Comment by tupui** [comment]
Thanks @drammock, I just merged it.

Yes for the formatting this is what I mean and what we see at the moment on the dev doc:

Screenshot 2024-01-04 at 22 34 49

Locally for me though, I have the correct content. We are hosting differently the doc for dev and releases and we have assumed it was an issue linked to how files are being distributed.

**Comment by tupui** [comment]
Ready from my side and I think we should merge ASAP. There are no new bugs, we finally can update and benefit from all the work that went through over the year. We can always work out extra CSS later on if wanted.

**Comment by drammock** [comment]
@tupui locally I've just fixed the left-sidebar not-in-a-list problem.  I'll open another PR to this branch.  I still can't grok why it's showing *the wrong TOC* though.

**Review by lucascolley — APPROVED** [review]

**Review by lucascolley — COMMENTED** [review]
~https://scipy.github.io/devdocs/dev/toolchain.html#building-the-documentation should be updated for the newrequirement on the theme.~ sorry for the noise, missed that this was already done!

**Comment by drammock** [comment]
The PR build is now doing something different: *no left-sidebar TOC at all* on the API pages: https://output.circle-artifacts.com/output/job/ddf5b596-d4b4-4ead-993a-697c7199f96d/artifacts/0/html/reference/index.html

I don't know all the details about how your stable vs dev docs are hosted, but @tupui seems to think it's OK if those sidebars don't look right? Not really my call to make. I'm happy to keep looking into this but can't promise I'll get to it today.

**Comment by tupui** [comment]
I just built locally and get the expected behaviour. So yes, I still think it's linked to the deployment. And I mean, even this new wrong behaviour would be better than the current wrong behaviour 😅 

Screenshot 2024-01-05 at 17 13 52


@drammock in an attempt to resolve the issue, I had made an attempt at re working the sidebar with #18604. But I did not get it to work. Maybe you would spot something I missed.

**Comment by 12rambau** [comment]
@tupui I checked the build of your documentation and I think you should consider moving the version swithcer back next to the logo of scipy. I checked on small screen and the navbar_end placeholder gets crowded, putting the titles on 2 lines: 

On the other hand you have lot of room left next to your logo: 
image

**Comment by tupui** [comment]
@12rambau happy to do that if that fixes the issue 😃 Do you want to make a PR on the branch or shall I do it?

**Comment by 12rambau** [comment]
as my screenshot shows I'm still on my company computer I should not be working on opensource 😄 so maybe for the sake of time it's better if you do it

**Comment by tupui** [comment]
Done 😉 It's slightly better. There is a small transition, but that's a nitpick at that point.

**Comment by j-bowhay** [comment]
Another nit but on small screens a burger button appears on the right which opens a menu with nothing on
image
image

**Comment by tylerjereddy** [comment]
Hey @tupui, did you want me to try to get this into `1.12.0` or `1.13.0`, based on how likely you think it is for problems to arise? `1.12.0` is almost at RC2 (waiting on a Pythran release I think); `1.13.0` might be closer to early February (to support NumPy `2.0.0`).

**Comment by tupui** [comment]
> Hey @tupui, did you want me to try to get this into 1.12.0 or 1.13.0, based on how likely you think it is for problems to arise? 1.12.0 is almost at RC2 (waiting on a Pythran release I think); 1.13.0 might be closer to early February (to support NumPy 2.0.0).

I would have tried 1.12.x if that's not so much trouble now. But your call. At that point we can wait a few more months if that's easier.

I would just like that we try to get this in quick. This way more people see the new style and people can improve/tweak the CSS if wanted.

> Another nit but on small screens a burger button appears on the right which opens a menu with nothing on

Yes this is to be done on the theme side. i.e. not for now. (IIRC I actually reported that at some point when I tested previous versions 😅)

**Comment by lucascolley** [comment]
> The PR build is now doing something different: no left-sidebar TOC at all on the API pages

I think that this could arguably block inclusion in 1.12.0 if a fix isn't found in time. The rest of the issues that have been raised seem minor though and should be fine to save for later IMO.

**Comment by tupui** [comment]
Please have a look at the current dev doc. It is actually worse. It shows the TOC from another section.

As I said above, we don't get the issue with a local build. We had a look with @melissawm a while back and could not get why this was the case. The positive thing we have seen though is that the once we release, all is good.

This could be explain as the deployment is different. Released docs are hosted on a server we administrate while dev and CI are both hosted on CircleCI and GitHub pages which both serve static files.

**Comment by tupui** [comment]
I found the issue for the TOC https://github.com/pydata/pydata-sphinx-theme/issues/854

@drammock this seems to still be an issue.

**Comment by lucascolley** [comment]
> Please have a look at the current dev doc. It is actually worse. It shows the TOC from another section.

Got you, sorry I was looking at the stable docs where it's okay. In that case, I don't see any blockers for this in 1.12 here! Of course, there could be other problems we haven't found yet, but I think the 'TOC from another section' is enough of a bug to justify the backport. And as you've said, this will unblock further improvements.

**Comment by drammock** [comment]
> I found the issue for the TOC [pydata/pydata-sphinx-theme#854](https://github.com/pydata/pydata-sphinx-theme/issues/854)
> 
> @drammock this seems to still be an issue.

That issue is about the *right* sidebar (the in-page TOC), I thought that was working ok here?

**Comment by tupui** [comment]
@drammock this is the issue: https://github.com/scipy/scipy/pull/16660#issuecomment-1879277743

**Comment by 12rambau** [comment]
for your landing page, could you add `html_theme.sidebar_secondary.remove: true` to your index.rst page metadata (very top of the page) 

This is what we are using for our landing page and it work like a charm.

File: doc/source/dev/toolchain.rst
Comment (h-vetinari) [inline doc/source/dev/toolchain.rst:419]: Can we update the dependencies here? For example, the new theme [requires](https://github.com/pydata/pydata-sphinx-theme/blob/v0.15.1/pyproject.toml#L23) at least sphinx 5. I'm surprised it has no lower bound on sphinx-design, but with the docs build [using](https://github.com/scipy/scipy/blob/7ff74ee275de307d6a17bebff4ea10cf58d35dff/.circleci/config.yml#L119) python 3.11, it [seems](https://github.com/executablebooks/sphinx-design/releases) we need at least 0.4.

And so on (matplotlib 2 is freaking ancient, should probably require 3.x for some recent x)

File: pyproject.toml
Comment (h-vetinari) [inline pyproject.toml:100]: Reflected here as well, obviously

**Review by h-vetinari — COMMENTED** [review]
Just a note that the theme version bump does not happen in isolation from the other components, and that we should update our deps/docs. Can happen in a follow-up PR though.

**Comment by tupui** [comment]
@h-vetinari I made the update 😃 

@drammock there is still an issue. I said earlier that locally the sidebar was ok for the reference pages, but I made a mistake and had 0.14.x when building. With 0.15.1, it's indeed empty. I played around and I can get the TOC back if I remove the hidden markup (see last commit). The output looks good, only issue is that the build time explodes again (see the CI before the last commit).

Hence now I pushed the version with the empty TOC as it's the only one that builds in a reasonable time.

I personally would be for going with this because it's still an improvement over the current state of the dev doc deployment. The drawback being that now to deploy releases, we would need to edit the `reference/index.rst` to remove `:hidden:`. (Unless someone finds another fix)

@tylerjereddy your call 😃 We can also modify `dev.py` to add a doc release. It would just do this small edit.

**Comment by drammock** [comment]
@tupui I think it's the same issue as https://github.com/pydata/pydata-sphinx-theme/issues/1625 and should hopefully be fixed by https://github.com/pydata/pydata-sphinx-theme/pull/1632. I'm planning to make some thorough tests on monday but if you can test that PR branch with scipy build that would be helpful

**Comment by drammock** [comment]
(also my best guess of what was wrong:  internally we had a `includehidden=False` default. When calculating the sidebar partial TOCTree it was thus ignoring any hidden TOCs, so for the API page the TOCTree got pruned down to only the non-hidden TOC --- AKA the development/contrib page one)

**Comment by h-vetinari** [comment]
> @h-vetinari I made the update 😃

Thank you, I appreciate it! Matplotlib is not 100% consistent; once you use `>=3.5` and once `>3.5`. I don't really care either way, JFYI.

Also, thanks a lot for your help @drammock & @12rambau! :pray:

**Comment by tupui** [comment]
@drammock I just tested and it does bring back the TOC on the reference page, but at the cost of a very long build time again.

**Comment by tylerjereddy** [comment]
Pythran is out so I may try to get RC2 out tomorrow. If this is to be backported, is someone confident enough to squash-merge and say that it is "somewhat safe?"

**Comment by drammock** [comment]
> Pythran is out so I may try to get RC2 out tomorrow. If this is to be backported, is someone confident enough to squash-merge and say that it is "somewhat safe?"

As an outsider I don't think I really get a vote here, but personally I'd prefer to wait. Weird things keep happening and I'd like to figure out why.

**Comment by tylerjereddy** [comment]
Understood, I'll leave the backport label but won't place extra strain here to get it done right away.

**Comment by drammock** [comment]
I've just opened another PR into @tupui's branch, thinking I had finally solved things. Then I build the docs locally one more time just to make sure, and discovered I was wrong, and I think now I might know why.  It appears that the API pages can *sometimes* end up with the wrong left-sidebar TOC *when building docs in parallel*.  I'm not yet sure how deterministic this is, so I'm doing a few more test builds to see if I can narrow it down.

**Comment by tupui** [comment]
Awesome thanks @drammock

**Comment by drammock** [comment]
So I'm pretty much convinced now that the "wrong sidebar on API pages" is 100% a problem of serial-vs-parallel sphinx build. When I build with `python dev.py doc -j1` it is always correct, when I build with `-j8` it is always wrong. So the short-term solution is to mark pydata-sphinx-theme as `parallel_write_safe=False` :(  until such time as I or someone else can figure out how to fix it.

On the plus side, with https://github.com/pydata/pydata-sphinx-theme/pull/1642 I now have the local SciPy doc build down to between 6 and 7 minutes with `-j1`. In addition, I think that once the theme is marked as parallel-write-unsafe, y'all could still use parallel jobs during *read* which should still yield a non-trivial speed up over single threaded.

I'll post back again once https://github.com/pydata/pydata-sphinx-theme/pull/1642 is merged, and once we've updated the `parallel_write_safe` flags for the theme.

**Comment by tupui** [comment]
Thanks @drammock for the investigation. 6-7 mins is definitely ok, is this with no-cache? That would be a good surprise as with the old version this is what I was getting but in parallel with `-j8`.

**Comment by drammock** [comment]
> Thanks @drammock for the investigation. 6-7 mins is definitely ok, is this with no-cache? That would be a good surprise as with the old version this is what I was getting but in parallel.

yes, this is with `--no-cache` :)

**Comment by drammock** [comment]
```console
$ python dev.py doc clean  # just to make sure
$ time python dev.py doc --no-cache -j8
...
WARNING: the pydata_sphinx_theme extension is not safe for parallel writing
WARNING: doing serial write
...
real    3m49.691s
user    10m5.980s
sys     2m42.991s
$ python dev.py doc clean  # just to make sure
$ time python dev.py doc --no-cache
...
real    8m1.070s
user    8m15.306s
sys     2m22.436s
```

(not sure why it's 8 min this time instead of 6-7 I saw before, but still an improvement!)

**Comment by tupui** [comment]
Nice! The scaling was either way not so linear 😅

**Comment by h-vetinari** [comment]
What's the status/plan here? Incidentally, I don't see any CI results anymore... 🤔

**Comment by lucascolley** [comment]
We're just waiting for https://github.com/pydata/pydata-sphinx-theme/pull/1642 I believe.

**Comment by drammock** [comment]
> We're just waiting for [pydata/pydata-sphinx-theme#1642](https://github.com/pydata/pydata-sphinx-theme/pull/1642) I believe.

I'll self-merge that this Friday unless one of the other maintainers asks for more time to review. Additional steps:

- merge https://github.com/tupui/scipy/pull/34 into this PR
- push out a point-release of the theme
- update the theme min-version constraint in your pyproject.toml / requirements.txt / CI configs / etc

**Comment by drammock** [comment]
@tupui we had a false start; the GH release of 0.15.2 didn't make it to PyPI due to a new upstream bug in sphinx causing the tests to fail (which prevented our `publish` action from succeeding).  But that's all fixed now, and 0.15.2 is now on PyPI. Can you manually re-trigger CIs here (or push an empty commit)?

**Comment by tupui** [comment]
Thanks @drammock, just re-launched 😃🤞

**Comment by melissawm** [comment]
13 mins! And it looks like the warnings are safe.

**Comment by tupui** [comment]
Just for fun, I clicked on the "Explain this error" with AI 🤣 Looks like Sphinx is still too hard for a LLM

Screenshot 2024-01-19 at 01 12 36

**Comment by tupui** [comment]
@drammock I will have to test locally but the sidebar is still off I am afraid. The build time is good though 😃 Was there another PR you wanted to make on this branch?

Screenshot 2024-01-19 at 01 23 17

**Comment by h-vetinari** [comment]
> I will have to test locally but online, the sidebar is still off I am afraid.

It's sidebar _and_ top bar, it seems. I.e. clicking on the "API reference" tab on the top still shows "Development" tab being highlighted (matching the content of the left sidebar).

**Comment by tupui** [comment]
Yes, same as current deployment.

**Comment by drammock** [comment]
Argh, ok. I'll try to look again tomorrow.

**Comment by drammock** [comment]
- [x] make sure that pydata-sphinx-theme 0.15.2 is actually getting installed into CircleCI environment (no "show installed versions" step in CircleCI, but the rendered doc says 0.15.2 in the footer, so we can be fairly sure)
- [ ] check if sphinx version matters.  CircleCI logs say it was built with Sphinx 5.3.0; my local tests from before were using Sphinx 7.2.6.
- [ ] ...
- [ ] profit

**Comment by tupui** [comment]
@drammock if building locally leads to all sidebars etc being ok. Then I think we are good. We can only know for sure once we deploy for real but it might just be again the same issue we have right now with our current dev doc.

**Comment by drammock** [comment]
> @drammock if building locally leads to all sidebars etc being ok. Then I think we are good. We can only know for sure once we deploy for real but it might just be again the same issue we have right now with our current dev doc.

in that case I guess to get the CircleCI build to "pass" you need to somehow make it OK/expected to see

```
WARNING: the pydata_sphinx_theme extension is not safe for parallel writing
WARNING: doing serial write
```

... or else stop running CircleCI with `-j2`

**Comment by tupui** [comment]
@drammock thank you. For now let's try with removing `-j2` because IIUC the warning is a log and would need to write a filter on an event handler.

Otherwise I had the chance to run this locally and it seems all good.

Screenshot 2024-01-20 at 00 50 08


@h-vetinari it would be helpful if you had time to confirm this too. And if everything is fine, I would suggest to move this forward.

**Comment by h-vetinari** [comment]
> @h-vetinari it would be helpful if you had time to confirm this too.

So far everything I've done was check the docs as rendered by CI (though IMO this is also the closest we're going to get in advance to what the actual deployment will look like), and with the latest change, the rendered docs look fine AFAICT.

Given the behaviour observed by @drammock, it seems pretty clear that the `-j2` is the culprit, given the lack of thread-safety of PST?

> And if everything is fine (and CI comes back green), I would suggest to move this forward.

Fine by me to move forward with this and merge, though given the current 42 commits[^1], I think we should either squash, or clean up the git history.

[^1]: the answer to life, the universe and everything!

**Comment by tupui** [comment]
@h-vetinari I am fine with a squash 👍

(In general I am a partisan of squash merge either way 😉)

**Comment by drammock** [comment]
Squash is fine with me, FWIW. I'll keep looking into why it wasn't enough to just mark the theme as parallel write unsafe... But that shouldn't hold up things here.

**Review by h-vetinari — APPROVED** [review]
Thanks for sticking through this slog @tupui, and for the help @drammock!

**Comment by tupui** [comment]
Thank you all for your help and dedication! This is great that we can finally move forward and be in sync with the theme.

@tylerjereddy would it be possible for you to re-publish the doc for 1.12 with this?

**Comment by tupui** [comment]
@h-vetinari Please feel free to merge. For the message I don't think we need much more than saying we updated the theme. No need to talk about all the commits in the squash. The only important thing to me is to keep the co-author mentions.

**Review by tylerjereddy — COMMENTED** [review]
`make dist` seems to succeed on this branch, so that's a good sign for the release process

If `1.13.0` comes out before `1.12.1` I don't know how much effort backporting is worth, but keeping the label for now seems "ok."

**Pull request merged**
